### ISSUE TITLE
[Question] RTC denoise update equation is not same with the paper

### ISSUE THREAD

**Issue by shlyang** [huggingface/lerobot#2511]
Hi, I'm recently trying to implement RTC based on [openpi](https://github.com/[Physical-Intelligence/openpi](https://github.com/Physical-Intelligence/openpi)).
I've noticed that in the paper and Physical-Intelligence's [real-time-chunking-kinetix](https://github.com/Physical-Intelligence/real-time-chunking-kinetix), the time goes from 0 to 1, however in [src/openpi/models/pi0.py](https://github.com/Physical-Intelligence/openpi/blob/3e0325a5e6032a9ab6beef7395fe47ca724980a5/src/openpi/models/pi0.py), the time goes from 1 to 0.
I agree with you that in the denoise_step, x1_t should be updated with  x_t - time * v_t, like [modeling_rtc.py#L215](https://github.com/huggingface/lerobot/blob/0f551df8f4bad4c504e395ea3df74fc5f714016f/src/lerobot/policies/rtc/modeling_rtc.py#L215)
But I'm confused at the v_t updating, in original paper, it should be 
<img width="851" height="69" alt="Image" src="https://github.com/user-attachments/assets/8d0bda33-1477-4ff3-ba6a-8f60d8228aee" />
so I update using `result = v_t + guidance_weight * correction`, which is opposite as yours code [modeling_rtc.py#L228](https://github.com/huggingface/lerobot/blob/0f551df8f4bad4c504e395ea3df74fc5f714016f/src/lerobot/policies/rtc/modeling_rtc.py#L228)
Can you tell me why you choose to update v_t with `result = v_t - guidance_weight * correction`?

**Comment by shlyang** [comment]
Here's my code
```python
    def realtime_action(
        self,
        rng: at.KeyArrayLike,
        obs: _model.Observation,  
        prev_action_chunk: at.Float[at.Array, "b ah ad"],
        *,
        num_steps: int,
        inference_delay: int,
        prefix_attention_horizon: int,
        prefix_attention_schedule: PrefixAttentionSchedule,
        max_guidance_weight: float,
    ) -> _model.Actions:
        """
        RTC(flow_matching)：t: 0→1
        x_t = (1 - time) * noise + time * action
        v_t = action - noise
       =>action = x_t + (1-t) * v_t
        
        pi0：t: 1→0
        x_t = time * noise + (1 - time) * actions
        v_t = noise - actions
        =>actions = x_t - t * v_t
        """
        observation = _model.preprocess_observation(None, obs, train=False)
        dt = -1.0 / num_steps
        batch_size = observation.state.shape[0]
        
        # reuse sample_actions()
        prefix_tokens, prefix_mask, prefix_ar_mask = self.embed_prefix(observation)
        prefix_attn_mask = make_attn_mask(prefix_mask, prefix_ar_mask)
        positions = jnp.cumsum(prefix_mask, axis=1) - 1
        _, kv_cache = self.PaliGemma.llm([prefix_tokens, None], mask=prefix_attn_mask, positions=positions)

        def step(carry, _):
            x_t, time = carry
            
            # @functools.partial(jax.vmap, in_axes=(0, 0, None))  # vmap over batch
            def pinv_corrected_velocity(x_t, y, t):
                def denoiser(x_t):
                    suffix_tokens, suffix_mask, suffix_ar_mask, adarms_cond = self.embed_suffix(
                        observation, x_t, jnp.full((batch_size,), t)
                    )
                    suffix_attn_mask = make_attn_mask(suffix_mask, suffix_ar_mask)
                    prefix_attn_mask_expanded = einops.repeat(prefix_mask, "b p -> b s p", s=suffix_tokens.shape[1])
                    full_attn_mask = jnp.concatenate([prefix_attn_mask_expanded, suffix_attn_mask], axis=-1)
                    positions = jnp.sum(prefix_mask, axis=-1)[:, None] + jnp.cumsum(suffix_mask, axis=-1) - 1
                    
                    # forward
                    (_, suffix_out), _ = self.PaliGemma.llm(
                        [None, suffix_tokens],
                        mask=full_attn_mask,
                        positions=positions,
                        kv_cache=kv_cache,
                        adarms_cond=[None, adarms_cond],
                    )
                    v_t = self.action_out_proj(suffix_out[:, -self.action_horizon :])[0]  # remove the batch dimension
                    result = x_t - t * v_t
                    return result, v_t
                
                x_0, vjp_fun, v_t = jax.vjp(denoiser, x_t, has_aux=True)
                weights = self.get_prefix_weights(
                    inference_delay, prefix_attention_horizon, self.action_horizon, prefix_attention_schedule
                )
                error = (y - x_0) * weights[None, :, None]
                pinv_correction = vjp_fun(error)[0]
                inv_r2 = (t**2 + (1 - t) ** 2) / (t ** 2)
                c = jnp.nan_to_num(t / (1 - t), posinf=max_guidance_weight)
                guidance_weight = jnp.minimum(c * inv_r2, max_guidance_weight)
                result = v_t + guidance_weight * pinv_correction
                return result
            
            v_t = pinv_corrected_velocity(x_t, prev_action_chunk, time)   
            return (x_t + dt * v_t, time + dt), None
        
        noise = jax.random.normal(rng, shape=(batch_size, self.action_horizon, self.action_dim))
        (x_0, _), _ = jax.lax.scan(step, (noise, 1.0), length=num_steps)
        return x_0
```

**Comment by helper2424** [comment]
Niice, I am doing the same at the moment, the jax version doesnt work well yet on my side.

**Comment by shlyang** [comment]
@helper2424 Can you explain why the v_t should be updated with '-'?

**Comment by helper2424** [comment]
In pi-like models the train target is subtraction of the velocity, not addition (it's opposite to the original FM). So, with the first case thats why `x_t - time * v_t`. 

With the second case, the logic is the following - we have an error function which is `err = (prev_chunk_left_over - x1_t) * weights`. And we want to calculate gradients of it based on previous `x_t` and correct the `v_t`  based on this gradients. Gradients usually point to the direction of function value increasing, so we want to move the opposite direction, decrease the error. In such case we should do subtraction. `v_new = v_t - correction * guidance`. This is my logic, @shlyang please correct me if I am wrong.

**Comment by shlyang** [comment]
Thanks for your reply. I originally thought this subtraction may be related to the opposite definition of denoising time, your explaination gives me a new perspective. Though I may do some experiments to verify this.

Another question, I'm afraid that executing `x_t.requires_grad_(True)` after `v_t = original_denoise_step_partial(x_t)` may result in `v_t` doesn't contain `x_t`'s grad, which makes the `correction` actually equals to `err`(∂x1_t/∂x_t=I). Maybe you can check this in debug_tracker?

**Comment by helper2424** [comment]
@shlyang Yes, sure, I will check.

**Comment by helper2424** [comment]
I have checked, yep with such approach the correction is equal to error. Good catch, thanks for checking! Probably I end up with it during the debugging process. With such approach the newly generated actions are perfectly aligned with previous actions chunks and we have smooth shift to new chunks, which kind of works. That was my target that I was expecting from the RTC. But it's a little bit far from the original paper. 

I have made a fix for that in the PR with additional improvements - https://github.com/huggingface/lerobot/pull/2499.

Also, I keep optional opportunity to use the current behavior, as it works with real robot. Probably the main reason why is because of RTC is not only in-painting but a big framework inference which consist of many pieces. And all of them bring it's own bonuses to the final result. Also, in real world the newly generated trajectories are not so far from the old one - so perfect alignment make the transition smooth, especially if we work with joint values. If trajectories will be very different from each other - than with current behavior trajectories is smooth, but arms could crazily accelerate during the shift from one to new one. It won't be done in one step, it will be smooth, but still big shift. If we take into the account model's gradients than with big differences between trajectories model won't align them well and probably will be shifting more to the new trajectory.

So, the PR is here - before merging need to make bunch of experiments. In any case the current approach aligns trajectories and works - so you can use it too :)

**Comment by shlyang** [comment]
Thanks, and this issue can be closed. :)

**Comment by helper2424** [comment]
@shlyang please share some insides or debug results if u would find anything

**Comment by shlyang** [comment]
okay

**Comment by shlyang** [comment]
I tried to use `error` substitute `correction`, got some interesting results. Below I ran inference on training data, colored lines are ground truth, the first is using `error` directly, the second use `correction`, the latter one can fit the gt well.

<img width="1500" height="500" alt="Image" src="https://github.com/user-attachments/assets/a4b07844-03ec-4ced-b9fb-5c8f58dc9162" />

<img width="1500" height="500" alt="Image" src="https://github.com/user-attachments/assets/2cc6adec-eb3f-4140-bab5-ff775d3f7d33" />

In fact, I calculated cosine simularity of `error` and `correction` in each denoising step, it varies between [-0.43, 0.92], and the 0.92 only achiedved at the 10th step, before that it's abs was rather small.
I used mae as metric and did sweep experiments, and the result is as below

<img width="2250" height="750" alt="Image" src="https://github.com/user-attachments/assets/f89608d8-2ae0-4f10-9663-98fe3b2e3dab" />
I tested infer_delay(1,2,3,4), use_subtraction(True/False), use_correction(True/False), use_mask(True/False) the x-axis item naming method follows  f"d{r['infer_delay']}_s{int(r['use_subtraction'])}_p{int(r['use_correction'])}_m{int(r['use_mask'])}"
currently my conclusion is, use_correction can obviously minimize the mae, use_subtraction is slightly better than add, and use_mask is nearly no affect.

Hope this can be helpful for you.

**Comment by helper2424** [comment]
Looks very interesting, can u help me to read these charts?

<img width="862" height="594" alt="Image" src="https://github.com/user-attachments/assets/f71566ae-7617-4d12-9713-11a29dd99ce4" />

Do I correctly understood that it's EE positions generated by some model, right? What does mean ground truth in your case - the previous chunk or the newly generated without RTC?

Regarding this one:

<img width="857" height="423" alt="Image" src="https://github.com/user-attachments/assets/36eeeb06-2277-4a3a-893e-a2be9e7d1aae" />

What does mean `use_mask`? 

Here is an important thing - the RTC generate a trajectory which is "kind of" in between old and new one. I was testing it mostly with Lerobot, with models which work with absolute joint values. The situation shouldn't be too different for deltas and EE data, but when u compare two trajectories need to be careful, as they could be not equal and that's ok.

**Comment by shlyang** [comment]
For the first two charts, the dash gray lines are predicted action chunks; they originally were joint values, and I converted them to EE position using my robot's URDF. The ground truth is from one episode of my teleoperation data, I finetuned pi05(with openpi repo)，and inference on train/eval data to see the finetune effect. In my case, the execute horizon is 10 and I predict new chunks every 10 steps, the circles means each chunk start step.

use_mask is a experiment set, my output action dim was smaller than 32, so I padded to match them, as you did in `rtc/modeling_rtc:195`,  I was afraid using the padded chunks to calculate correction may import some wrong gradients and finally got unexpected new `v_t`, but the results showed the `use_mask` bring no big impact, so you don't need to pay attention to it.

Further more, I did a new sweep experiment, here's the results.

<img width="1590" height="955" alt="Image" src="https://github.com/user-attachments/assets/5b4ef14c-e0c0-43eb-94c5-5d3ee9eb3843" />

when I use `error` to update `v_t`, the mae is quite huge, but using subtraction may reduce the impacts, this may  indirectly confirms why your previous results were good.
when I only keep the 'use_pinv == True' rows, the results were cleaner. But `num_steps == 5` ones had lower mae, which was against my intuition, I began to suspect if my implemetation has some bugs.

<img width="1589" height="955" alt="Image" src="https://github.com/user-attachments/assets/899d9194-9b96-412f-aed9-ad0a16fec9bd" />

**Comment by helper2424** [comment]
Ok, got it, you use your openpi implementation and their repo 👍
